main.tf

```diff
@@ -25,7 +29,10 @@ locals {
 resource "aws_vpc" "this" {
   count = var.create_vpc ? 1 : 0
 
-  cidr_block                       = var.cidr
+  cidr_block          = local.cidr
+  ipv4_ipam_pool_id   = local.ipv4_using_ipam ? var.ipv4_ipam_pool_id : null
+  ipv4_netmask_length = local.ipv4_ipam_with_explicit_netmask
```

Comment:
I'm not following the logic of the locals but based on the API of the Terraform resource I think these should just simply be

```hcl
  cidr_block  					= var.ipv4_ipam_pool_id != null ? null : var.cidr_block
  ipv4_ipam_pool_id		= var.ipv4_ipam_pool_id
	ipv4_netmask_length = var.ipv4_netmask_length
```

Reply:
This resource has nuanced logic and is more complex. Heres a few lines that are important to consider and ill attempt to convey some of the nuances:
- cidr_block can now be explict or [derived from ipam](https://github.com/hashicorp/terraform-provider-aws/blob/8dfef413a46c1a22e49cf0af7b67a3b902746698/internal/service/ec2/vpc.go#L46)
- cidr_block `ConflictsWith` `ipv4_netmask_length` [link](https://github.com/hashicorp/terraform-provider-aws/blob/8dfef413a46c1a22e49cf0af7b67a3b902746698/internal/service/ec2/vpc.go#L77)

because of these hard schema requirements, you can pass attributes in the following configuration:


```terraform
module "no_ipam_vpc_example" {
  source = "../.."
  name   = "no-ipam-${local.name}"
  cidr   = "172.2.0.32/28"
}

module "ipv4_ipam_explicit_cidr_vpc" {
  source            = "../.."
  name              = "ipv4-explicit-cidr-${local.name}"
  ipv4_ipam_pool_id = aws_vpc_ipam_pool.ipv4_example.id
  cidr              = "172.2.0.32/28"
}

module "ipv4_ipam_explicit_netmask_vpc" {
  source              = "../.."
  name                = "ipv4-explicit-netmask-${local.name}"
  ipv4_ipam_pool_id   = aws_vpc_ipam_pool.ipv4_example.id
  ipv4_netmask_length = 28
}

module "ipv4_ipam_default_netmask_vpc" {
  source            = "../.."
  name              = "ipv4-default-netmask-${local.name}"
  ipv4_ipam_pool_id = aws_vpc_ipam_pool.ipv4_example.id
}
```

With this logic we can catch them during compile rather than at runtime

Reply:
you're correct. been a long couple of days! 😅  thanks for taking a closer look at that

Reply:
not quite 1 for 1 but this felt reminiscent: https://www.theolognion.com/senior-developer-successfully-converted-12-lines-of-bash-script-into-complex-microservice-architecture/ 

🤣